subsys/net/lib/aws_jobs/src/aws_jobs.c

```diff
@@ -0,0 +1,318 @@
+ *
+ * @param[in] ret Return value of snpritnf.
+ * @param[in] size Max size of what would have been the final string.
+ * @param[in] entry Name of the function it was called from.
+ */
+static inline void report_snprintf_err(u32_t ret, u32_t size, char *entry)
+{
+	if (ret >= size) {
+		LOG_ERR("Unable to allocate memory for %s", entry);
+	} else if (ret < 0) {
+		LOG_ERR("Output error for %s was encountered with return value"
+			" %d", entry, ret);
+	}
+}
+
+
+/**
+ * @brief Constructs the get topic for accepted and rejeceted AWS jobs.
+ *
+ * @param[in] client_id Client ID of the current connected MQTT session.
+ */
+static struct mqtt_topic construct_get_topic(u8_t *client_id)
+{
+	char get_topic[GET_TOPIC_LEN + 1];
+	int ret = snprintf(get_topic,
+			   GET_TOPIC_LEN,
+			   GET_TOPIC_TEMPLATE,
+			   client_id,
+			   "#");
+
+	report_snprintf_err(ret, GET_TOPIC_LEN, "get_topic");
+
+	struct mqtt_topic topic = {
+		.topic = {
+			.utf8 = get_topic,
+			.size = strlen(get_topic)
+		},
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE
+	};
+
+	return topic;
+}
+
+
+int aws_jobs_subscribe_get(struct mqtt_client *client)
+{
+
+
+	struct mqtt_topic subscribe_topic = construct_get_topic(
+						client->client_id.utf8
+						);
+	const struct mqtt_subscription_list subscription_list = {
+		.list = (struct mqtt_topic *)&subscribe_topic,
+		.list_count = 1,
+		.message_id = SUBSCRIBE_GET
+	};
+	LOG_DBG("Subscribe: %s\n", subscribe_topic.topic.utf8);
+
+	return mqtt_subscribe(client, &subscription_list);
+}
+
+
+/**
+ * @brief Constructs the notify-next topic for reciving AWS IoT jobs.
+ *
+ * @param[in] client_id Client ID of the current connected MQTT session.
+ */
+static struct mqtt_topic construct_notify_next_topic(u8_t *client_id)
+{
+	char notify_next_topic[NOTIFY_NEXT_TOPIC_MAX_LEN  + 1];
+	int ret = snprintf(notify_next_topic,
+			   NOTIFY_NEXT_TOPIC_MAX_LEN,
+			   NOTIFY_NEXT_TOPIC_TEMPLATE,
+			   client_id);
+	report_snprintf_err(ret,
+			    NOTIFY_NEXT_TOPIC_MAX_LEN,
+			    "notify_next_topic");
+
+	struct mqtt_topic topic = {
+		.topic = {
+			.utf8 = notify_next_topic,
+			.size = strlen(notify_next_topic)
+		},
+		.qos = MQTT_QOS_1_AT_LEAST_ONCE
+	};
+
+	return topic;
+}
+
+
+int aws_jobs_subscribe_notify_next(struct mqtt_client *client)
+{
+
+
+	struct mqtt_topic subscribe_topic = construct_notify_next_topic(
+						client->client_id.utf8
+						);
+	const struct mqtt_subscription_list subscription_list = {
+		.list = (struct mqtt_topic *)&subscribe_topic,
+		.list_count = 1,
+		.message_id = SUBSCRIBE_NOTIFY_NEXT
+	};
+	LOG_DBG("Subscribe: %s\n", subscribe_topic.topic.utf8);
+
+	return mqtt_subscribe(client, &subscription_list);
+}
+
+
+/**
+ * @brief Constructs the notifyt topic for reciving AWS IoT jobs.
+ *
+ * @param[in] client_id Client ID of the current connected MQTT session.
+ *
+ * @retval mqtt_topic Returns formated mqtt_topic.
+ */
+static struct mqtt_topic construct_notify_topic(u8_t *client_id)
```

Comment:
`char` is a bit misleading as it can be `utf-8`

Reply:
I think it makes sense because we assume it's null-terminated. But I see that `u8_t` is the convention in `mqtt.h`. Did you consider passing a `struct mqtt_utf8*` instead?

Reply:
I don't mind either, maybe it's good to keep convention from `mqtt.h` or maybe not. As you said we do expect a string which is null terminated as `snprintf` will probably fail if it gets something unexpected. So I don't think passing in `struct mqtt_utf8` would really give any benefits as we won't be able to use the size anyway.